kernel/kernel-api/src/main/java/io/delta/kernel/Transaction.java

```diff
@@ -0,0 +1,207 @@
+                    ColumnarBatch data = filteredColumnarBatch.getData();
+                    for (String partitionColName : partitionValues.keySet()) {
+                        int partitionColIndex = data.getSchema().indexOf(partitionColName);
+                        data = data.withDeletedColumnAt(partitionColIndex);
+                    }
+                    return new FilteredColumnarBatch(data,
+                            filteredColumnarBatch.getSelectionVector());
+                }
+        );
+    }
+
+    /**
+     * Get the context for writing data into a table. The context tells the connector where the data
+     * should be written and what should be the target file size.
+     *
+     * @param tableClient      {@link TableClient} instance to use.
+     * @param transactionState The transaction state
+     * @param partitionValues  The partition values for the data. If the table is un-partitioned,
+     *                         the map should be empty
+     * @return {@link DataWriteContext} containing metadata about where and how the data for
+     * partition should be written.
+     */
+    static DataWriteContext getWriteContext(
+            TableClient tableClient,
+            Row transactionState,
+            Map<String, Literal> partitionValues) {
+        validatePartitionValues(
+                getPartitionColumnsList(transactionState),
+                partitionValues);
+
+        String targetDirectory = getTargetDirectory(
+                getTableRoot(transactionState),
+                getPartitionColumnsList(transactionState),
+                partitionValues);
+        return new DataWriteContext(
+                targetDirectory,
+                getTargetFileSize(transactionState),
+                partitionValues,
+                getStatisticsColumns(tableClient, transactionState));
+    }
+
+    /**
+     * Stage the given data files as a Delta actions that can be committed in a transaction.
+     *
+     * @param tableClient      {@link TableClient} instance.
+     * @param transactionState State of the transaction.
+     * @param fileStatusIter   Iterator of row objects representing each data file written.
+     * @param dataWriteContext The context used when writing the data files given in
+     *                         {@code fileStatusIter}
+     * @return
+     */
+    static CloseableIterator<Row> stageAppendOnlyData(
+            TableClient tableClient,
+            Row transactionState,
+            CloseableIterator<DataFileStatus> fileStatusIter,
+            DataWriteContext dataWriteContext) {
+        URI tableRoot = new Path(getTableRoot(transactionState)).toUri();
+        return fileStatusIter.map(
+                fileWriteResult -> {
+                    Path filePath = new Path(fileWriteResult.getPath());
+                    Map<Integer, Object> valueMap = new HashMap<>();
+                    valueMap.put(0, tryRelativizePath(filePath, tableRoot).toString()); // path
```

Comment:
question: can we make 0, 1, 2, 3, 4 etc be enums or static final values for the indexes? would that be less error prone than just in line magic numbers? doesn't need to be done in this PR

Reply:
Moved this to a utility method in `AddFile`.